### PR TITLE
feat: Security screen

### PR DESCRIPTION
**Pull request by pedrohusky**
## Description 
     - You setup a password in .env.
     - When bot loads, if there is any password set, it will show the security window. It's window can't be closed, moved, and content behind it cannot be touched.

## Motivation and Context

Some people were asking, then I did it. Is is simple, but seems strong at the password verification code.

## How Has This Been Tested?

Setting a password, reading the password.

I tried to bypass the security system without success. I did a good job. :)

## Screenshot

Captura de Tela (151)

@edit

Sorry my other pull request was wrong.

### CODE REVIEW

File: public/js/App.js [hunk 1]
```diff
@@ -229,6 +231,14 @@ class App extends React.Component {
                 dustTransfer={dustTransfer}
                 sendWebSocket={this.sendWebSocket}
               />
+              {password != {} ? (
+                <PasswordScreen
+                  password={password}
+                  jsonStrings={languageData}
```
Comment (pedrohusky) [inline public/js/App.js:237]: Sorry this shouldn't be here.

File: app/frontend/websocket/handlers/latest.js [hunk 2]
```diff
@@ -44,7 +48,8 @@ const handleLatest = async (logger, ws, _payload) => {
       accountInfo: JSON.parse(cacheTrailingTradeCommon['account-info']),
       exchangeSymbols: JSON.parse(cacheTrailingTradeCommon['exchange-symbols']),
       publicURL: cacheTrailingTradeCommon['local-tunnel-url'],
-      apiInfo: binance.client.getInfo()
+      apiInfo: binance.client.getInfo(),
+      password: passArray
```
Comment (chrisleekr) [inline app/frontend/websocket/handlers/latest.js:52]: This is not gonna work as WebSocket value is public.

Comment (pedrohusky) [reply app/frontend/websocket/handlers/latest.js:52]: I didn't understand why it won't gonna work.

**Comment by chrisleekr** [comment]
@pedrohusky 

The way you developed the feature is not gonna work because the password is exposed to the public.
You will be able to see the password in the Inspection -> Web Socket.

What I intended with securing the frontend is as below:

1. Set default password in the configuration
2. Provide two options to protect the frontend
    - Protect symbols & settings: When access to the frontend, it should display the password field and authenticate to see the symbols and settings.
    - Protect settings only: When access to the frontend, it displays the authentication. When try to change the settings, it should show the password field and authenticate.
3. Authentication must be done in the backend, not in the frontend. Once authentication is done, each WebSocket request must contain an authentication header and the backend to verify prior to executing the command.

I hope it makes sense.

**Comment by pedrohusky** [comment]
> @pedrohusky 
> 
> The way you developed the feature is not gonna work because the password is exposed to the public.
> You will be able to see the password in the Inspection -> Web Socket.

Thanks for the heads up. I don't understand much about that. Learning everyday :)
> 
> What I intended with securing the frontend is as below:
> 
> 1. Set default password in the configuration
> 2. Provide two options to protect the frontend

It would be two options that could be selected?

>     - Protect symbols & settings: When access to the frontend, it should display the password field and authenticate to see the symbols and settings.

Maybe a blur behind the modal? :)

>     - Protect settings only: When access to the frontend, it displays the authentication. When trying to change the settings, it should show the password field and authenticate.

I don't know if secure only settings is a useful feature as someone could still manual trade your symbols. Or they would be secured too?

> 3. Authentication must be done in the backend, not in the frontend. Once authentication is done, each WebSocket request must contain an authentication header and the backend to verify prior to executing the command.

Then I will need to call a new file like the ones called by configure.js, to verify the password there, right?

> 4. In the settings, you should be able to change the password. 
>

Of course.

**Comment by chrisleekr** [comment]
> Thanks for the heads up. I don't understand much about that. Learning everyday :)

Haha, me too.

> It would be two options that could be selected?

That should be in the configuration and can be updated in the UI.

> Maybe a blur behind the modal? :)

Not actually. The Web Socket should not send any data unless it's authenticated. Again, it is exposed to the public.

> I don't know if secure only settings is a useful feature as someone could still manual trade your symbols. Or they would be secured too?

Oh, sure sure. The trading feature should be authenticated as well.

> Then I will need to call a new file like the ones called by configure.js, to verify the password there, right?

I haven't thought or plan how to do it yet as this was a lower priority. 
But it must be thoroughly tested.

**Comment by pedrohusky** [comment]
@chrisleekr 

I did some changes. Now on WebSocket the only public info is:

```
"login": {
          "passwordActivated": false - To see if there is a need to create the security screen, otherwise show the bot right away
          "logged": false, -  A boolean telling if you're logged in or not.
          "elapsedTime": 0, - The last date you logged in to calculate with below variable the difference in minutes between them
          "loginWindowMinutes": 60 - The default login window time. It will have an option at frontend with change password to change this number
        }
```

I'm finishing the options fields.

**Comment by pedrohusky** [comment]
@chrisleekr 

I'm not understanding this conflict between the branches...

**Comment by pedrohusky** [comment]
@chrisleekr Hi.

You need to accept this pull request as sooner as possible. I found a groundbreaking bug.

My bot sent me this link: https://default.loca.lt

When I enter it, it isn't my version of the bot, the coins monitored are not what I'm monitoring and I can see the balance of the account and make any actions I want.

I didn't made any action because I'm a good person. Someone could not be. 

@chrisleekr please see it.

**Comment by chrisleekr** [comment]
@pedrohusky 

Hmm, that trader should not use the default local tunnel URL. I didn't imagine someone can get `default` local tunnel.
API key/secret are secured as the bot will not expose them to the public.
The trader must change the default local tunnel URL.

It looks like the trader used the default setting.
I will make the default setting to not use the local tunnel. 

I won't just merge the PR before thoroughly tested and match with the requirements.

**Comment by pedrohusky** [comment]
> @pedrohusky 
> 
> Hmm, that trader should not use the default local tunnel URL. I didn't imagine someone can get `default` local tunnel.
> API key/secret are secured as the bot will not expose them to the public.
> The trader must change the default local tunnel URL.
> 
> It looks like the trader used the default setting.
> I will make the default setting to not use the local tunnel. 
> 
> I won't just merge the PR before thoroughly tested and match with the requirements.
> 
> 
> 
> 
> 

Hi. It is not one trader. I just entered 4 different with the same url.



My bot sent me the default link today. At 9am, and didn't change it until I restarted it.
My link is now a correct one, like 'dragonfruit-husky'

It seems even with the tunnel set right, sometimes your bot will show you that link. Exposing to others.

The above screenshot, the trader is using a version of your bot before the profit/loss accordion.

**Comment by chrisleekr** [comment]
The local tunnel will configure the subdomain like `dragonfruit-husky` when it couldn't acquire the requested subdomain due to the subdomain is unavailable.

If you set `default` subdomain, then the bot continuously attempts to acquire that subdomain. 
You shouldn't use `default` as the local tunnel subdomain.

**Comment by pedrohusky** [comment]
> If you set `default` subdomain, then the bot continuously attempts to acquire that subdomain. 
> You shouldn't use `default` as the local tunnel subdomain. 

But why it is default at your .env.dist file and there is no step in readme warning us of this?

Mine is default from when I first downloaded the bot. I never changed it. Never had to.

**Comment by chrisleekr** [comment]
@pedrohusky 

Yap, you are right with it. Hence, I just committed the env and warning in the README.md

- https://github.com/chrisleekr/binance-trading-bot/blob/master/.env.dist
- https://github.com/chrisleekr/binance-trading-bot#how-to-use

Thanks for reporting.

**Comment by chrisleekr** [comment]
Hi @pedrohusky 

I have pushed the code for securing the frontend.
Thanks to your code, I got a lot of inspiration.

I will close this PR in favour of #260.

**Pull request closed without merging**